File: Balanced Binary Tree/Balanced Binary Tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:

    int height(TreeNode* node){
        if(!node)   return 0;
        int left_height = 0;
        int right_height = 0;
        if(node->left)      left_height = height(node->left);
        if(node->right)     right_height = height(node->right);
        return (left_height > right_height) ? left_height + 1 : right_height + 1;
    }

    bool isBalanced(TreeNode *root) {
        if(!root)       return true;
        int height_diff = height(root->left) - height(root->right);
        if((height_diff > 1) || (height_diff < -1)){
            // Left and right subtrees have heights with a difference greater than 1
            return false;
        }
        return isBalanced(root->left) && isBalanced(root->right);
    }
};

File: Balanced Binary Tree/Balanced Binary Tree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    
    private int depth(TreeNode node){
        if(node == null){
            return 0;
        }
        else{
            return Math.max(depth(node.left), depth(node.right)) + 1;
        }
    }
    
    public boolean isBalanced(TreeNode root) {
        if(root == null){
            return true;
        }
        else{
            return Math.abs(depth(root.left) - depth(root.right)) > 1 ? false : isBalanced(root.left) && isBalanced(root.right);
        }
    }
}
